### PR TITLE
T2D-223 Web resource accepts an empty URL

### PR DESCRIPTION
**Pull request by iyangar**
Fixes

### CODE REVIEW

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 1]
```diff
@@ -305,6 +305,15 @@ private String postTestWebResource() throws Exception {
         return mvcResult.getResponse().getHeader("Location");
     }
 
+    @Test
+    public void postWebResourceNullURL() throws Exception {
```
Comment (cyenyxe) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: Could you please add another test that fails for a non-empty, but still invalid URL?

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java [hunk 2]
```diff
@@ -61,6 +63,8 @@
     @URL
     @JsonProperty
     @Column(nullable = false, columnDefinition = "TEXT")
+    @NotEmpty
+    @Pattern(message = "Must be a valid URL.", regexp="^(https?|ftp|file):(//|\\\\)[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]")
```
[the remark this answers is not in the document]
Comment (iyangar) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java]: Sure, i have removed "file" protocol and its gets blocked also. But existing @URL code bars no-protocol URLs. If we remove @URL we will loose JSR validations for URL.

[the remark this answers is not in the document]
Comment (Zhicheng-Liu) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java]: Ah, I forgot there is a `@URL` annotation. It seems that `@URL` validation can take a few optional settings, like `protocol` and `regexp`: https://docs.jboss.org/hibernate/validator/5.4/api/org/hibernate/validator/constraints/URL.html

If these work for our purpose, I would suggest to use these instead of the `@Pattern` annotation.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java [hunk 3]
```diff
@@ -58,9 +61,10 @@
 
     @ApiModelProperty(position = 3, required = true)
     @NotNull
-    @URL
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java]: Could you please also removed the `import` for this?

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java [hunk 4]
```diff
@@ -29,6 +30,8 @@
 import javax.persistence.GenerationType;
 import javax.persistence.Id;
 import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+import java.lang.annotation.ElementType;
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java]: The `ElementType` is not used anywhere. It should not be left here.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java [hunk 5]
```diff
@@ -58,9 +61,10 @@
 
     @ApiModelProperty(position = 3, required = true)
     @NotNull
-    @URL
     @JsonProperty
     @Column(nullable = false, columnDefinition = "TEXT")
+    @NotEmpty
+    @Pattern(message = "Must be a valid URL.", regexp="(^(https?|ftp):)?(//|\\\\)[-a-zA-Z0-9+&@#/%?=~_|!:,.;$'*\\[\\]()]*")
```
Comment (Zhicheng-Liu) [inline src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java]: The `regexp` is wrong. It will match `//www.ebi.ac.uk`.

File: src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java (hunk 3 above)
Comment (iyangar) [reply src/main/java/uk/ac/ebi/ampt2d/metadata/persistence/entities/WebResource.java]: done

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 6]
```diff
@@ -305,6 +305,22 @@ private String postTestWebResource() throws Exception {
...
 
+    @Test
+    public void postWebResourceInvalidURL() throws Exception {
+        WebResource testWebResource = new WebResource(WebResource.Type.CENTER_WEB, "");
+
+        mockMvc.perform(post("/webResources")
+                .content(testWebResourceJson.write(testWebResource).getJson()))
+                .andExpect(status().is4xxClientError());
+
+        testWebResource = new WebResource(WebResource.Type.CENTER_WEB, "https://www.ebi.ac.uk<>");
```
Comment (cyenyxe) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: Could you please add some more tests to check valid and invalid URLs? For instance, making sure that only HTTP, HTTPS and FTP protocols are considered valid, as mentioned during @Zhicheng-Liu's review. Testing arbitrary errors in the URL would be also a good addition.

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 7]
```diff
@@ -305,6 +333,81 @@ private String postTestWebResource() throws Exception {
         return mvcResult.getResponse().getHeader("Location");
     }
 
+    private String postTestHTTPWebResource() throws Exception {
```
Comment (Zhicheng-Liu) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: You can group these four "repeated" `private` functions `postTest*WebResource` into one function taking one argument for the URLs with different protocols. Even further, you can also perform the `get` and `andExpect` checks in the same function.

File: src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java [hunk 8]
```diff
@@ -305,6 +333,81 @@ private String postTestWebResource() throws Exception {
...
+                .content(testWebResourceJson.write(testWebResource).getJson()))
+                .andExpect(status().is4xxClientError());
+
+        testWebResource = new WebResource(WebResource.Type.CENTER_WEB, "//fileserver/code/src/main/app.java");
+
+        mockMvc.perform(post("/webResources")
+                .content(testWebResourceJson.write(testWebResource).getJson()))
+                .andExpect(status().is4xxClientError());
+
+
```
Comment (Zhicheng-Liu) [inline src/test/java/uk/ac/ebi/ampt2d/metadata/MetadataApplicationTest.java]: Personally I would not leave that many blank lines here. There is a coding style guide here for your reference: https://www.ebi.ac.uk/seqdb/confluence/display/VAR/Java+coding+style
